### PR TITLE
New short-read assembler: PenguiN

### PR DESCRIPTION
**Pull request by danejo3**
PenguiN (Protein-guided Nucleotide Assembler) is a relatively new assembler that takes a slightly different approach to assembly. It uses the coding sequences of reads and translates them into protein sequences to guide the assembly process. According to the creator of PenguiN, "The main purpose of PenguiN is the assembly of complex metagenomic and metatranscriptomic datasets."

In our own tests, we have found that the final contigs may introduce unintended SNPs, depending on the number of iterations provided to the assembler. To address this issue, we clean up the contigs using `pilon` to polish the reads. This tool has shown a significant improvement in base accuracy.

One of the benefits we have found when using PenguiN is its ability to provide multiple full-length contigs. While this could be seen as a downside, as opposed to having a single consensus sequence, it gives users the option to choose the best contig avaliable.

Notes:
- One useful `extra_args` entry that users can input into YEAT's config is `--num-iterations`. (The test results mentioned above used 5 (default), 10, and 15.)
- No Bandage files are created for PenguiN assemblies because no `.gfa` or similar files are created by the assembler.

### CODE REVIEW

File: environment.yml [hunk 1]
```diff
@@ -25,5 +25,9 @@ dependencies:
     - nanofilt>=2.3
     - nanoplot>=1.20
     - hifiasm>=0.19
-    - hifiasm_meta>=hamtv0.3
+    - hifiasm_meta=hamtv0.2
```
Comment (danejo3) [inline environment.yml:28]: `hifiasm_meta` tests running version `hamtv0.3` are failing due to the lack of memory and "complexity of the problem" on my 64 gb of RAM laptop. When reverting back down to `hamtv0.2`, these tests pass normally. As of now, pinning the version number to keep `hifiasm_meta` tests running without crashing.

File: yeat/config/assembly.py [hunk 2]
```diff
@@ -93,7 +93,7 @@ def get_target_files(self):
     def get_qa_file(self, sample):
         readtype = self.get_readtype(sample)
         algorithm_dir = f"analysis/{sample.label}/{readtype}/{self.label}/{self.algorithm}"
-        if self.bandage:
+        if self.bandage and self.algorithm != "penguin":
             return f"{algorithm_dir}/bandage/.done"
         return f"{algorithm_dir}/quast/report.html"
```
Comment (danejo3) [inline yeat/config/assembly.py:98]: No assembly graph instructions (such as .gfa files) are provided by PenguiN. Skipping this step if `Bandage` is called.

File: yeat/workflow/Paired [hunk 3]
```diff
@@ -58,3 +58,23 @@ rule unicycler:
...
+    input:
+        read1="seq/downsample/{sample}/paired/{sample}.R1.fq.gz",
+        read2="seq/downsample/{sample}/paired/{sample}.R2.fq.gz"
+    threads: 128
+    params:
+        outdir="analysis/{sample}/paired/{label}/penguin",
+        extra_args=lambda wildcards: config["assemblies"][wildcards.label].extra_args
+    shell:
+        """
+        penguin guided_nuclassemble {input.read1} {input.read2} {params.outdir}/unpolished_contigs.fasta {params.outdir} --threads {threads} {params.extra_args}
```
Comment (danejo3) [inline yeat/workflow/Paired:75]: PenguiN is primarily used for short-reads only. No support for long read assembly. Possibly could attempt to assembly long reads by using the single-end mode with PenguiN. I have not tested this personally.

File: yeat/workflow/Paired [hunk 4]
```diff
@@ -58,3 +58,23 @@ rule unicycler:
...
+        bowtie2-build {params.outdir}/unpolished_contigs.fasta {params.outdir}/unpolished_contigs.fasta
+        bowtie2 -p {threads} -x {params.outdir}/unpolished_contigs.fasta -1 {input.read1} -2 {input.read2} 2> {params.outdir}/unpolished_contigs.bowtie.log | samtools view -b -@ {threads} | samtools sort -@ {threads} -o {params.outdir}/unpolished_contigs.sorted.bam
+        samtools index {params.outdir}/unpolished_contigs.sorted.bam
+        pilon --genome {params.outdir}/unpolished_contigs.fasta --bam {params.outdir}/unpolished_contigs.sorted.bam --output {params.outdir}/contigs
```
Comment (danejo3) [inline yeat/workflow/Paired:79]: Pilon requires a `bam` file of the mapped reads in order to polish the contigs. We use `bowtie2` and `samtools` to map and index the reads.

**Review by danejo3 — COMMENTED** [review]
This PR is ready for review! @standage

**Review by standage — APPROVED** [review]
Looks great!

**Pull request merged**